src/resource.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python3
+# Copyright 2021 Martin Kalcok
+# See LICENSE file for licensing details.
+#
+# Learn more at: https://juju.is/docs/sdk
+
+"""Abstraction of Kubernetes Resources.
+
+This module contains classes that wrap kubernetes resources and enable unified
+interface for their removal.
+"""
+# Remove this pylint skip when functionality is added to `Resource` class
+# pylint: disable=too-few-public-methods
+from typing import Callable
+
+from kubernetes.client import AppsV1Api, CoreV1Api
+from kubernetes.client import RbacAuthorizationV1Api as RbacAuthApi
+from kubernetes.client import StorageV1Api
+
+
+class MissingMethod(BaseException):
+    """Exception that represents method that is not implemented."""
+
+
+class Resource:
+    """Base class for Kubernetes resources.
+
+    So far, main purpose of this class is to provide unified `remove()` method
+    that calls appropriate k8s api method.
+    """
+
+    def __init__(self, name: str, namespace: str = ""):
+        """
+        Initialize k8s resource with name and optionally namespace.
+
+        If namespace is not provided, resource is treated as cluster-wide.
+        :param name: resource name
+        :param namespace: resource namespace
+        """
+        self.name = name
+        self.namespace = namespace
+
+    @property
+    def _remove_action(self) -> Callable:
+        """Return method of k8s api client that removes cluster resource."""
+        raise MissingMethod(
+            "Removal of {} {} is not " "implemented".format(self.__class__.__name__, self.name)
+        )
+
+    @property
+    def _remove_namespaced_action(self) -> Callable:
+        """Return method of k8s api client that removes namespaced resource."""
+        raise MissingMethod(
+            "Removal of namespaced {} {} is not "
+            "implemented".format(self.__class__.__name__, self.name)
+        )
+
+    def remove(self) -> None:
+        """Call appropriate api method to remove k8s resource."""
+        if self.namespace:
+            self._remove_namespaced_action(self.name, self.namespace)
+        else:
+            self._remove_action(self.name)
+
+
+class CoreResource(Resource):
```

Comment:
Similar to the previous comment, should this class be named `KubernetesCoreResource` ?